src/libutil/args.cc

```diff
@@ -69,6 +76,36 @@ void Args::parseCmdline(const Strings & _cmdline)
     }
 
     bool argsSeen = false;
+
+    // Heuristic to see if we're invoked as a shebang script, namely,
+    // if we have at least one argument, it's the name of an
+    // executable file, and it starts with "#!".
+    Strings savedArgs;
+    auto isNixCommand = std::regex_search(programName, std::regex("nix$"));
+    if (isNixCommand && cmdline.size() > 0) {
+        auto script = *cmdline.begin();
+        try {
+            auto lines = tokenizeString<Strings>(readFile(script), "\n");
+            if (std::regex_search(lines.front(), std::regex("^#!"))) {
+                lines.pop_front();
+                for (auto pos = std::next(cmdline.begin()); pos != cmdline.end();pos++)
+                    savedArgs.push_back(*pos);
+                cmdline.clear();
+
+                for (auto line : lines) {
+                    line = chomp(line);
+
+                    std::smatch match;
+                    if (std::regex_match(line, match, std::regex("^#!\\s*nix\\s(.*)$")))
+                        for (const auto & word : shellwords(match[1].str()))
+                            cmdline.push_back(word);
```

Comment:
@roberth i considered an “else, break out of scanning” condition here.

Reply:
What would you think about doing a regex check for `^[^#\/\\\*-]` and breaking out?

Reply:
That might be a bit aggressive.
The Rust ones here are already quite interesting https://nixos.wiki/wiki/Nix-shell_shebang
I think I've seen a `stack run` + `nix-shell` one once, which was pretty weird.
Your suggestion does happen to work, but I can't help feeling that it's by coincidence.
Perhaps breaking out on the first empty line would be sensible enough? I think those only tend to occur after the shebang lines. I could be wrong, and it'd be an easy mistake to add a newline between two tools' shebang stuff. I guess Nix would be the first one, so it all works out?

Worst case we can add extra instructions to let the script configure shebang scanning later.